operationsgateway_api/src/records/export_handler.py

```diff
@@ -0,0 +1,428 @@
+        created file(s) to the zip file being created ready for download.
+        """
+        try:
+            # first check that there should be a waveform to process
+            _ = record_data["channels"][channel_name]
+        except KeyError:
+            # there is no entry for this channel in the record
+            # so there is no waveform to process
+            return
+
+        if self.export_waveform_csvs or self.export_waveform_images:
+            log.info(
+                "Getting waveform to add to zip: %s %s",
+                record_id,
+                channel_name,
+            )
+            try:
+                waveform_model = Waveform.get_waveform(
+                    Waveform.get_relative_path(record_id, channel_name),
+                )
+            except Exception:
+                self.errors_file_in_memory.write(
+                    f"Could not find waveform for {record_id} {channel_name}\n",
+                )
+                # no point trying to process the waveform so return at this point
+                return
+
+            if self.export_waveform_csvs:
+                num_points = len(waveform_model.x)
+                waveform_csv_in_memory = io.StringIO()
+                for point_num in range(num_points):
+                    waveform_csv_in_memory.write(
+                        str(waveform_model.x[point_num])
+                        + ","
+                        + str(waveform_model.y[point_num])
+                        + "\n",
+                    )
+                self.zip_file.writestr(
+                    f"{record_id}_{channel_name}.csv",
+                    waveform_csv_in_memory.getvalue(),
+                )
+                ExportHandler.check_zip_file_size(self.zip_file_in_memory)
+
+            if self.export_waveform_images:
+                # if rendered trace images have been requested then add those
+                waveform = Waveform(waveform_model)
+                waveform_png_bytes = waveform.get_fullsize_png(x_units, y_units)
+                self.zip_file.writestr(
+                    f"{record_id}_{channel_name}.png",
+                    waveform_png_bytes,
+                )
+                ExportHandler.check_zip_file_size(self.zip_file_in_memory)
+
+    def _add_main_csv_file_to_zip(self):
+        """
+        If other files have been exported and therefore a zip file is being prepared
+        for export then add the CSV file to the zip file.
+        """
+        if self.export_scalars:
+            if (
+                len(self.zip_file.infolist()) > 0
+                and len(self.main_csv_file_in_memory.getvalue()) > 0
+            ):
+                self.zip_file.writestr(
+                    f"{self.get_filename_stem()}.csv",
+                    self.main_csv_file_in_memory.getvalue(),
+                )
+                ExportHandler.check_zip_file_size(self.zip_file_in_memory)
+
+    def _add_messages_file_to_zip(self):
```

Comment:
Could some logging be added to log out what errors are being written to the file please? When I was fixing the failing test due to merging the echo waveforms changes in, the output of the test mentioned the `EXPORT_ERRORS.txt` file. But as I didn't have access to the zip (as a service manager might not in production), it was difficult to work out what the errors were.

Reply:
Good idea! Fixed in https://github.com/ral-facilities/operationsgateway-api/pull/104/commits/c071491d22a949d773ff3ea965b7c5c632bc75e8